### PR TITLE
feat: contains

### PR DESCRIPTION
**Pull request by ThePrimeagen**
A couple of helpers to use with node comparison.  This is particularly
useful when using selections and node querying or scope determination.

### CODE REVIEW

**Comment by ThePrimeagen** [comment]
I just realized there's a small bug.  I'll fix that shortly. By the way is this something that would fit within this repo?

**Comment by theHamsta** [comment]
>  By the way is this something that would fit within this repo?

Yeah, absolutely. Maybe we have to align the naming with the other functions and either overload `in_node_range` (to make it work with pos,range or node) or clarify through naming that the one function is for a position, the other for a range, and another for a node `node_contains_pos`,  `node_contains_range`, `node_contains_node`.

**Comment by ThePrimeagen** [comment]
If anyone can tell me what to name it, i'll rename it!  Also, I need to run the linting.  Let me know what to rename it to and I'll do that.

Also should end / start be inclusive?

**Comment by ThePrimeagen** [comment]
@theHamsta I also had to update the range match algo. it turns out I wasn't necessarily thinking about it complete enough.  The algo is now "correct" as far as I can tell

**Comment by ThePrimeagen** [comment]
So as I do more research I think that this may be extra work that isn't needed.

```lua
-- I need to experiment with this
LanguageTree:contains({self}, {range})               *LanguageTree:contains()*
                Determines wether This goes down the tree to recursively check childs.

                Parameters: ~
                    {range}  is contained in this language tree

                Parameters: ~
                    {range}  A range, that is a `{ start_line, start_col,
                             end_line, end_col }` table.
                    {self}
```

**Comment by theHamsta** [comment]
I don't think your functions are related with `LanguageTree:contains` just checks the sub-ranges of a language.

File: lua/nvim-treesitter/ts_utils.lua [hunk 1]
```diff
@@ -5,6 +5,57 @@ local utils = require "nvim-treesitter.utils"
...
+-- @param b the node to be contained
+function M.node_contains(a, b)
+  if a == nil or b == nil then
+    return false
+  end
+
+  local start_row, start_col, end_row, end_col = b:range()
+  return M.is_in_node_range(a, start_row, start_col) and
+    M.is_in_node_range(a, end_row, end_col)
+end
```
Comment (theHamsta) [inline lua/nvim-treesitter/ts_utils.lua:19]: I don't think this is correct since the end position is exclusive Otherwise this function would also be more elegant for the other one. Maybe have a helper function to obtain the char right before `end_row`,`end_col`? 

 It should also be possible to have only one implementation for those two functions. They are two function with completely different names doing the exact same thing. `M,get_node_range` coerces nodes and ranges to ranges.

Comment (theHamsta) [reply lua/nvim-treesitter/ts_utils.lua:19]: But, wait in the second function you're using vim ranges. Then, the two functions are actually not the same and the one with start_row/col end_row/col is actually inclusive while the node one is not

**Review by theHamsta — CHANGES_REQUESTED** [review]
So I think this is ok to merge after you checked whether the node-in-node function handles the last char correct (the function should be symmetrical between a and b).

File: lua/nvim-treesitter/ts_utils.lua [hunk 2]
```diff
@@ -5,6 +5,57 @@ local utils = require "nvim-treesitter.utils"
...
+
+-- determines if a node exists within a range.  Imagine a range selection
+-- across '<,'> and an identifier.  Does the identifier exist within the
+-- selection?
+--
+-- @param a the containing node
+-- @param b the node to be contained
+M.range_contains_node = function(node, start_row, start_col, end_row, end_col)
```
Comment (theHamsta) [inline lua/nvim-treesitter/ts_utils.lua:27]: The documentation is not up-to-date

**Comment by ThePrimeagen** [comment]
@theHamsta Just to be clear

1.  Update documentation
2.  What do you mean symmetrical for node_contains?  I assume this means the following cases. 
  a.  If a == b, it returns true for both
  b.  If a contains b is true, b contains a will return false     

Also, does nvim-treesitter have plenary tests?  If you want to avoid the dependency, there could always be a separate repo (though that would be super sucky)

File: lua/nvim-treesitter/ts_utils.lua [hunk 3]
```diff
@@ -5,6 +5,57 @@ local utils = require "nvim-treesitter.utils"
...
+    return true
+  elseif start_row == end_row then
+    return node_start_row == start_row and node_end_row == end_row and
+    start_col <= node_start_col and end_col >= node_end_col
+  elseif start_row == node_start_row and start_row == node_end_row then
+    return start_col <= node_start_col
+  elseif end_row == node_start_row and end_row == node_end_row then
+    return end_col >= node_end_col
+  elseif start_row <= node_start_row and end_row >= node_end_row then
+    return start_col <= node_start_col and end_col >= node_end_col
```
Comment (kyazdani42) [inline lua/nvim-treesitter/ts_utils.lua:53]: just a hint here, you don't need to elseif after a return, a simple `if` works (and is usually easier to read :) )

**Comment by theHamsta** [comment]
1)  Update documentation
2) Make sure all the functions you wrote are correct
  - I guess `node_contains` is not correct. The end point of a node is not in its own range
3) Make sure code you add doesn't add confusion
   - `is_in_node_range` sounds like it could fit for positions and nodes. Or you make it explicit :
             - `node_contains_pos`/`node_contains_node`/`visual_selection_contains_node`, or 
             - `pos_is_in_node_range`/`node_is_in_node_range`/`node_is_in_visual_selection`
             -  Maybe `node_is_in_visual_selection` could live in the `incremental_selection` module to reduce conflicts with the similarly named functions in this module
   - Vim ranges vs tree-sitter ranges always add confusion. Maybe the term `visual_selection` instead of `range` would prevent to accidentally use `node:range()` with this function
4) Make sure that your PR does not add code duplication.
   - `is_in_node_range` and `range_contains_node` do the same dangerous multicase check but slightly differently since `range_contains_node` uses Vim ranges. Make sure it is used in the visual_section module when this is the use case. Because there a similar complicated check is performed. It was incorrect in multiple ways in the beginning which only became clear in edge cases
5) nvim-treesitter has plenary (see scripts/run_tests.sh). At the moment it's not in CI because indent tests are failing. So you would make sure that only the new test you would add get run on CI or you only run the test locally

There are multiple ways how this can be solved I sure you'll find a good solution. Please understand that I'm so pedantic about this row/col/range topic but we had a lot of pain with the Vim vs treesitter ranges (with zero vs one-based with inclusive vs exclusive).

File: lua/nvim-treesitter/ts_utils.lua (hunk 3 above)
Comment (ThePrimeagen) [reply lua/nvim-treesitter/ts_utils.lua:53]: ah yes.  The original version of this had some elseif checks that would fall through to the bottom.  I think that is an astute point that as this has changed I no longer need the `elseif`.  And honestly, I would prefer not having them.

**Comment by ThePrimeagen** [comment]
> Please understand that I'm so pedantic about this row/col/range topic but we had a lot of pain with the Vim vs treesitter ranges (with zero vs one-based with inclusive vs exclusive).

Trust me, you are not being pedantic, you are keeping things on rails.  1/2 of my time spent on playing around with tree sitter is spent thinking about the type of range / inclusive vs exclusive.  Its a real pain and I appreciate your thoroughness.  My goal isn't to "get a sweet commit and t shirt" but to make an operation I am doing 50 times all over in an upstream lib where many people can benefit from this.

> There are multiple ways how this can be solved I sure you'll find a good solution. 
You clearly think to high of my skallz

I'll address your points likely today/tomorrow

**Comment by theHamsta** [comment]
> Trust me, you are not being pedantic, you are keeping things on rails. 1/2 of my time spent on playing around with tree sitter is spent thinking about the type of range / inclusive vs exclusive. Its a real pain and I appreciate your thoroughness. My goal isn't to "get a sweet commit and t shirt" but to make an operation I am doing 50 times all over in an upstream lib where many people can benefit from this.

Something that helped me is to just always use tree-sitter ranges and positions `local range = {node:range()}` as a 4 tuple `range[1], range[2] ...` instead of col/row, col/row and have them the only kind of range that we have in this repo. They are consistent with LSP ranges and positions (zero-base, exclusive). It helps to convert the Vim values to tree-sitter ones before you do any logic on them.

A useful utility function could calculate `the char before that one` which is also often needed to get from exclusive to the last char in the range

**Comment by theHamsta** [comment]
There is also a function `get_node_range` which coerces 4-tuple ranges and nodes into 4-tuple ranges.

**Comment by ThePrimeagen** [comment]
Telescopic Johnson and I have been batting around the idea of a meta table that just does it all for you.

**Comment by theHamsta** [comment]
Like TSRange in this repo that is node&range at the same time? Just bring it to upstream nvim then it's handy for everyone ;-)

**Comment by ThePrimeagen** [comment]
hasn't fallen off radar, just got caught up :)

**Comment by clason** [comment]
superseded by rewrite on `main` branch (which removes `ts_utils`; sufficiently common functionality should be added upstream, otherwise just implement this locally -- we could have a "common snippets" wiki page as a compromise)
